File: core/src/Prediction.cpp
#include "cmaclib/Prediction.h"

#include <cmaclib/Result.h>
#include <utility>

using namespace CmacLib;

void Prediction::SetValues(std::vector<double>&& values)
{
	this->values = ::std::move(values);
}

void Prediction::SetActiveWeightIndices(std::vector<unsigned int>&& activeWeightIndices)
{
	this->activeWeightIndices = ::std::move(activeWeightIndices);
}

void Prediction::SetActiveWeights(std::vector<std::vector<double>>&& weights)
{
	this->activeWeights = ::std::move(weights);
}

void Prediction::SetBasisValues(std::vector<double>&& basis)
{
	this->basisValues = ::std::move(basis);
}

Prediction::Prediction()
	: values(std::vector<double>())
	, activeWeightIndices(std::vector<unsigned int>())
	, activeWeights(std::vector<std::vector<double>>())
	, basisValues(std::vector<double>()){}

Prediction::~Prediction(){}

const std::vector<double>& Prediction::GetValues()
{
	return this->values;
}

const std::vector<unsigned int>& Prediction::GetActiveWeightIndices()
{
	return this->activeWeightIndices;
}

const std::vector<std::vector<double>>& Prediction::GetActiveWeights()
{
	return this->activeWeights;
}

const std::vector<double>& Prediction::GetBasisValues()
{
	return this->basisValues;
}

/// @brief Get the calculated multiplication and accumulation
/// of basis and active weights.
/// @param pPrediction The pointer to the prediction instance.
/// @return Cmac output.
CDoubleArray CCmacLibIPredictionGetValues(
	CCmacLibIPrediction* pPrediction
) {
	return {
		(double*) pPrediction->GetValues().data(),
		pPrediction->GetValues().size()
	};
}

/// @brief Get the indices of the active weights.
/// @param pPrediction The pointer to the prediction instance.
/// @return Active weight indices.
CUintArray CCmacLibIPredictionGetActiveWeightIndices(
	CCmacLibIPrediction* pPrediction
) {
	return {
		(unsigned int*) pPrediction->GetActiveWeightIndices().data(),
		pPrediction->GetActiveWeightIndices().size()
	};
}

/// @brief Get the values of the active weights.
/// @param pPrediction The pointer to the prediction instance.
/// @return Active weight values.
/// (Remember to delete the CDoubleArray data pointer.)
CDoubleArrayArray CCmacLibIPredictionGetActiveWeights(
	CCmacLibIPrediction* pPrediction
) {
	/// @brief The container for the active weights 2-D array.
	CDoubleArrayArray matrix;
	matrix.size = pPrediction->GetActiveWeights().size();
	matrix.data = new CDoubleArray[matrix.size];

	for (unsigned int i = 0; i < pPrediction->GetActiveWeights().size(); i++) {
		matrix.data[i].data = (double*) pPrediction->GetActiveWeights().data()->data();
	}

	return matrix;
}

/// @brief Get the basis function values.
/// @param pPrediction The pointer to the prediction instance.
/// @return Basis function values array. (No need to deallocate).
CDoubleArray CCmacLibIPredictionGetBasisValues(
	CCmacLibIPrediction* pPrediction
) {
	
	return {
		(double*) pPrediction->GetBasisValues().data(),
		pPrediction->GetBasisValues().size()
	};
}

/// @brief Indicates if an operation was successful.
/// @param pPrediction The pointer to the prediction instance.
/// @return Success.
cmaclib_bool_t CCmacLibIPredictionIsSuccessful(
	CCmacLibIPrediction* pPrediction
) {
	return pPrediction->IsSuccessful();
}

/// @brief If not successful, a failure message is provided.
/// @param pPrediction The pointer to the prediction instance.
const char* CCmacLibIPredictionGetMessage(
	CCmacLibIPrediction* pPrediction
) {
	return pPrediction->GetMessage().c_str();
}

/// @brief Destructor.
/// @param ppPrediction The pointer to the pointer to the prediction instance.
void CCmacLibIPredictionDestroy(
	CCmacLibIPrediction** ppPrediction
) {
	if (*ppPrediction) {
		delete *ppPrediction;
		*ppPrediction = nullptr;
	}
}

File: wrappers/julia/src/prediction.jl
#!/usr/bin/julia

# Prediction type declaration.
export IPrediction
"""
Prediction output.
"""
struct IPrediction end

export IPredictionGetValues
"""
Get the calculated multiplication and accumulation
of basis and active weights.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
Cmac output.
"""
function IPredictionGetValues(pInst::Ptr{IPrediction})::Vector{Cdouble}
    c_data::CDoubleArray = ccall(
        (:CCmacLibIPredictionGetValues, cmaclib_core),
        CDoubleArray, (Ptr{IPrediction}, ), pInst)
    return unsafe_wrap(Vector{Cdouble}, c_data.data, c_data.size)
end

export IPredictionGetActiveWeightIndices
"""
Get the indices of the active weights.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
Active weight indices.
"""
function IPredictionGetActiveWeightIndices(pInst::Ptr{IPrediction})::Vector{Cuint}
    c_data::CUintArray = ccall(
        (:CCmacLibIPredictionGetActiveWeightIndices, cmaclib_core),
        CUintArray, (Ptr{IPrediction}, ), pInst)
    return unsafe_wrap(Vector{Cuint}, c_data.data, c_data.size)
end

export IPredictionGetActiveWeights
"""
Get the values of the active weights.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
Active weight values.
"""
function IPredictionGetActiveWeights(pInst::Ptr{IPrediction})::Vector{Vector{Cdouble}}
    c_data = ccall((:CCmacLibIPredictionGetActiveWeights, cmaclib_core),
        CDoubleArrayArray, (Ptr{IPrediction}, ), pInst)

    # The container for the active weights to be returned.
    matrix::Vector{Vector{Cdouble}} = []
    c_matrix::Vector{CDoubleArray} = unsafe_wrap(Vector{CDoubleArray},
        c_data.data, c_data.size)

    for col_vec_wrapped::CDoubleArray in c_matrix
        col_vec::Vector{Cdouble} = unsafe_wrap(
            Vector{Cdouble}, col_vec_wrapped.data, col_vec_wrapped.size
        )
        push!(matrix, col_vec)
    end

    # Properly Dispose c_data.
    ccall((:CDoubleArrayArrayDestroy, cmaclib_core),
        Cvoid, (CDoubleArrayArray, ), c_data)

    return matrix
end

export IPredictionGetBasisValues
"""
Get the basis function values.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
Basis function values.
"""
function IPredictionGetBasisValues(pInst::Ptr{IPrediction})::Vector{Cdouble}
    c_data::CDoubleArray = ccall((:CCmacLibIPredictionGetBasisValues, cmaclib_core),
        CDoubleArray, (Ptr{IPrediction}, ), pInst)
    return unsafe_wrap(Vector{Cdouble}, c_data.data, c_data.size)
end

export IPredictionIsSuccessful
"""
Indicates if an operation was successful.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
Success.
"""
function IPredictionIsSuccessful(pInst::Ptr{IPrediction})::Bool
    return ccall((:CCmacLibIPredictionIsSuccessful, cmaclib_core),
        Bool, (Ptr{IPrediction}, ), pInst)
end

export IPredictionGetMessage
"""
If not successful, a failure message is provided.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
## Returns
The failure message.
"""
function IPredictionGetMessage(pInst::Ptr{IPrediction})::Cstring
    return ccall((:CCmacLibIPredictionGetMessage, cmaclib_core),
        Cstring, (Ptr{IPrediction}, ), pInst)
end

export IPredictionDestroy
"""
Destructor.

## Arguments
- `pInst::Ptr{IPrediction}`: The pointer to the prediction instance.
"""
function IPredictionDestroy(pInst::Ptr{IPrediction})::Nothing
    ccall((:CCmacLibIPredictionDestroy, cmaclib_core),
        Cvoid, (Ptr{IPrediction}, ), pInst)
end